Documentation for “Secure ILIAS” (#1969)
ILIAS-eLearning/ILIAS: docs/configuration/secure.md

# Hardening and Security Guidance

This guideline is going to show you some best practice examples how to improve the security of your ILIAS installation, of the webserver and it's associated components.
The ILIAS e.V. requested a documentation for a "Secure ILIAS" in march 2019.
The section "Hardening and Security Guidance" should be removed and the security related instructions have to be mantained in a own document.

## Firewall

Block all traffic by default and explicitly allow only specific traffic to port 443/TCP for HTTPS secured traffic.
For "quality of life", it is recommend to also permit 80/TCP for a [redirect to HTTPS](#redirect-all-unencrypted-traffic-to-https-1).

## File Access Rights

If you're an experienced admin you MAY want to use more restrictive file access rights that we RECOMMEND in this document. To make it impossible for an attacker to modify PHP files if he gains control over the web server processes, those files SHOULD be owned by ```root``` wherever possible.

The only files and directories that MUST be owned/writeable by the web user are:

  * ilias.ini.php
  * data/
  * ILIAS data dir outside of the webservers docroot

> All the other files and directories should be owned by ```root```, but readable by the web user (e.g. 644/755).

Reviewer: Having the files not writeable, ok, but owned by root? If you want to make sure (beside the missing write permission) that the webserver could not alter the files make a new user/group or use "nobody". But "root" isn't a good advice (accidentally set setuid flags could even help an attacker to get full access) and would lead to problems if more than one admin with limited access rights should handle the installation.

Reviewer 2: Also it is enough to make the files readable by the web server, and not just everyone.
As the web servers process owner usually comes with its own group, say user ``www-data`` has a group ``www-data`` with itsels as the only member, a ``chown -R root:www-data``  allows to keep the permissions set to ``640/750``. (Replace `root` by `nobody` or whoever you choose as owner of the files.)

Author: In the updated PR, the username is changed to a placeholder (specific) user.

example of the suggested configuration:
```
git clone --single-branch -b release_5-4 https://github.com/ILIAS-eLearning/ILIAS.git \ /var/www/ilias/.

mkdir -p /var/www/ilias;

> mkdir /var/www/ilias_data;

Reviewer 3: I think we should use a common path for the "directory outside web root". The installation documentation uses `/var/www/files` as an example. Even though I'm not completely happy with this choice too (I consider `/var/www` as the default web root), we should at least be consistent. That makes it easier for new users/admins.

Reviewer 2: I think ``/var/lib/ilias`` would be the FHS-standard location for "data outside the web root" and I'd like to propose this location instead.

Reviewer 3: I would agree with `/var/lib/ilias` as the default location (even though we use `/var/iliasdata` at our company). The exact path is not really important, as long as these requirements are met:
1. It has to be outside the document root of the webserver (not under `/var/www`)
2. It should be somewhere under `/var`
3. It should be consistent across all documentation (especially the installation documentation)

Author: The documentation has to be a detailed step-by-step instruction, but the final path can be choiceable of the admin.

we add:

> example of the suggested configuration (may you can choose other directories, based on your discretion / distribution standard ):

to:

> example of the **suggested** configuration

So as the discussion in workshop, we only give suggestion.

Reviewer 4: I think the proposal  /var/www/ilias_data is very dangerous. Especially for sysadmins who have little knowledge. It could happen very quickly that the entire data directory becomes public. Why don't you use the variant as suggested by @ClemensRobbenhaar: /var/lib/ilias OR /var/share/ilias - wathever - but not /var/www?

> mkdir /var/www/ilias_log;

Reviewer 2: Same here: ``/var/log/ilias`` seems to me the FHS standard place instead.

Reviewer 3: I agree with that too. This is the path a system administrator unfamiliar with ILIAS would look for log files.

Author: I'ved change the path an fix an issue related to open_basedir.

mkdir /var/www/ilias_log/ilias_errors

> chown [www-data:www-data|www_ilias_1:www_ilias_1] /var/www/ilias /var/www/ilias/data

Reviewer 3: I don't understand this syntax. Is it an OR-clause (user/group `www-data` OR `www_ilias_1`)? The user/group `www_ilias_1` has not been introduced before in this document. Seems rather confusing to me.

Author: In workshop we discus to add a "list of placeholder names"  and use predefined/described placeholder.

chown [www-data:www-data|www_ilias_1:www_ilias_1] /var/www/ilias_data
chown [www-data:www-data|www_ilias_1:www_ilias_1] /var/www/ilias_log
chown [www-data:www-data|www_ilias_1:www_ilias_1] /var/www/ilias_log/ilias_errors

> chmod [2]775 /var/www/ilias/data

Reviewer 3: Like above, I don't quite understand the syntax here. Also, I'm not sure if the setgid bit should be set here. What is the reasoning behind it?

Author: That generated new files or directory will get also the group.
(e.g. may if there will be generate new clients)

chmod [2]775 /var/www/ilias_data
chmod [2]775 /var/www/ilias_log
chmod [2]775 /var/www/ilias_log/ilias_errors
```

After the installation of ILIAS is finished, you SHOULD also revoke write permission for the file ```ilias.ini.php``` (e.g. ```/var/www/ilias/ilias.ini.php```).

> note: for changing base setting via ILIAS setup, you need to grant write permission for the file ```ilias.ini.php``` again.

Reviewer 3: I'm not sure if we can recommend this. If the file is not writeable, the setup will throw errors and the root cause may not be obvious. I suspect that this note is easily forgotten after a while.

Reviewer 2: I really recommend making the ``ilias.ini.php`` read-only - after all it is not about an "easy to admins" install, but a secure install.

## Place data directory outside of the web root

It is highly RECOMMENDED to place your data directory outside of the web server docroot, as pointed out by the ILIAS Installation Wizard.

## Isolate Docroot
You MAY use openbasedir-restriction to avoid malicious software to directory-traverse out of your docroot-directory. This is very important if there are other websites on the same host.

Apache2:

>     php_admin_value open_basedir ./:%docroot_dir%:/usr/share/php/:/var/www/lib/:/tmp

Reviewer 3: Is `%docroot_dir%` a default (environment) variable? I don't find any documentation on this.

Reviewer 2: I think for apache2 it is called ``%{DOCUMENT_ROOT}`` - see  https://httpd.apache.org/docs/2.4/expr.html#vars under "Other request related variables".

Author: We define a placeholder list, so this is obsolete.

Nginx:

    fastcgi_param PHP_VALUE open_basedir="./:%docroot_dir%:/usr/share/php/:/var/www/lib/:/tmp";
**Hint:** This option will be applied to the PHP-FPM process. If there are multiple websites on your webserver you have to define a single PHP-FPM-pool for each website. Otherwise these other homepages won' t be accessible anymore.

## OS user handling security

> If you use PHP-FPM (FastCGI Process Manager), you can increase security by running the PHP-FPM processes as an specific instead ```www-data``` / ```wwwrun``` (depends on linux distribution).

Reviewer 3: Wording. Did you mean "[...]as a specific user instead of[...]"?

Author: thanks!

Snippet from a PHP-FPM pool definition:
```
; Unix user/group of processes
; Note: The user is mandatory. If the group is not set, the default user's group
;       will be used.
user = www_ilias_1

> group = www_ilias_1

Reviewer 3: User/group `www_ilias_1` was never introduced before in the document. It may not be clear that this user/group has to be created first.

Author: --> placeholder list

```

**note:**  

> NGINX and also apache2 is can only run with one user (no multi user multi process model).  

Reviewer 2: Grammar: "is can only" -> "can only"  ;)

Author: thanks!

So it is necessary to put all the "PHP-FPM"-users in the primary group of the webserver user.

## Major security improvement: use HTTPS

You can get a trusted, free SSL certificate at https://letsencrypt.org.
Or use a SSL certificate from a commercial certificate authority.

### Redirect all unencrypted traffic to HTTPS

To redirect all HTTP traffic to HTTPS you SHOULD issue a permanent redirect using the 301 status code:

#### Apache2

```
<VirtualHost *:80>
   ServerName yourservername.org
   Redirect permanent / https://yourservername.org/
</VirtualHost>
```

#### NGINX

```
server {
		listen   *:80;
		listen   [::]:80;
		server_name yourservername.org;

		location / {
				return 301 https://yourservername.org$request_uri;				
		}
```

### Proper SSL configuration

The default SSL settings (e.g. ciphers & ssl protocol version ) used by web servers are often not state-of-the-art, so you SHOULD consider to choose your own settings.

> **Suggestion of modern settings for SSL configuration:**

Reviewer: Instead of a static recommendation we just might should advice the user to use https://ssl-config.mozilla.org to get a more suited config for his environment.

Reviewer 3: Absolutely. These recommendations will be out-of-date in no time.

Author: There is a link to "Mozilla SSL Configuration Generator" in the following text.

We add a sentence that is necessary to revise this setting.  

The workshop discus that they have to be a suggested configuration for new admins.
We promise to hold this settings modern in future in this document.

    ssl_prefer_server_ciphers on;

    ssl_dhparam /etc/ssl/private/dhparam.pem;

```

``ssl_dhparam /etc/ssl/private/dhparam.pem;``:

This specifies a file with DH parameters for EDH (Ephemeral Diffie-Hellman) ciphers.  
By default, NGINX will use the default DHE paramaters provided by openssl. This uses a weak key that gets lower scores.  
Run ``openssl dhparam -out /etc/ssl/private/dhparam.pem 4096`` in terminal to generate it.

We RECOMMEND to use the [Mozilla SSL Configuration Generator](https://ssl-config.mozilla.org/) to generate a suitable configuration and the [Qualys SSL Labs Tests](https://www.ssllabs.com/ssltest/) or the [High-Tech Bridge SSL Server Test](https://www.htbridge.com/ssl/) to check your settings. It is recommended, to reach a "A" rating as minimum.

### Serve security related Headers

To improve the security of your ILIAS users you SHOULD set the following Headers:

```
    X-Content-Type-Options: nosniff;
    X-XSS-Protection: 1; mode=block;
    X-Frame-Options: SAMEORIGIN;
    Referrer-Policy: strict-origin;
    Feature-Policy sync-xhr 'self';  

>     add_header Content-Security-Policy "default-src 'self'; connect-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval' blob:; style-src 'self' 'unsafe-inline' data:; img-src 'self' 'unsafe-inline' data:; font-src 'self' 'unsafe-inline' data:; media-src 'self' 'unsafe-inline' data:";  

Reviewer 3: I am not aware of a Content-Security-Policy that does not lead to problems with ILIAS. Are you sure we can recommend this? Further down you say that this works "in most use cases". How extensively has this been tested?

Author: This results from many professional installation in several companies and organization.
So we can say, this is "tested".

>     add_header Content-Security-Policy "default-src 'self'; connect-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval' blob:; style-src 'self' 'unsafe-inline' data:; img-src 'self' 'unsafe-inline' data:; font-src 'self' 'unsafe-inline' data:; media-src 'self' 'unsafe-inline' data:";  

Reviewer: I wouldn't make this CSP recommendation for the same reasons as corro. A link to https://observatory.mozilla.org should be enough. If we really want to use CSP we should avoid the need for "unsafe-inline" completely, otherwise it doesn't make much sense.

Author: With the requirement for the secure document it is necessary to give admins a detailed configuration suggestion.

As a result of discussion, we provide a more or sometimes less "best practice" configuration.

         # Working with ILIAS in most use cases:
         add_header Content-Security-Policy "default-src 'self'; connect-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval' blob:; style-src 'self' 'unsafe-inline' data:; img-src 'self' 'unsafe-inline' data:; font-src 'self' 'unsafe-inline' data:; media-src 'self' 'unsafe-inline' data:";
         #compatibility to  Microsoft Internet Explorer 10:
         add_header X-Content-Security-Policy "default-src 'self'; connect-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval' blob:; style-src 'self' 'unsafe-inline' data:; img-src 'self' 'unsafe-inline' data:; font-src 'self' 'unsafe-inline' data:; media-src 'self' 'unsafe-inline' data:";
 ```

note:  
If you use a proxied [Chat Server](https://github.com/ILIAS-eLearning/ILIAS/blob/release_5-4/Modules/Chatroom/README.md), you MUST add the url to the CSP definition:  
```connect-src 'self' wss://onscreenchat.yourservername.org https://onscreenchat.yourservername.org;```

#### Validate the header configuration
It is recommended to validate your configuration with the services from https://securityheaders.com/.  
Try to reach **A** grade.

> ### Enable OCSP stapling (TLS Certificate Status Request)

Reviewer: OCSP stapling is more a privacy than a security feature. Maybe we should just link to some external documentation here as the prerequirements for OCSP stapling may vary a lot (some environments might force you to download your own OCSP files, don't have access to DNS, etc.).

Author: It will also improve the performance and also add more privacy since certificates are checked for actual validity or if they are revoked/compromised.
So we can provide this detailed and not only less by a reference.

> The  TLS/SSL extension [OCSP stapling](https://en.wikipedia.org/wiki/OCSP_stapling) provides a way to improve the performance (cause the webserver is will done this instead of the browser client) of SSL negotiation while maintaining visitor privacy.

Reviewer 3: Wording: "[...]the webserver is will done this[...]", did you mean "[...]the webserver will do this[...]"?

Author: yes, thanks.

The [Online Certificate Status Protocol (OCSP)](https://en.wikipedia.org/wiki/Online_Certificate_Status_Protocol) is used for checking the revocation status of a SSL certificate (X.509 certificate). -> https://tools.ietf.org/html/rfc6066

By adding the following to your SSL VirtualHost configuration your webserver will do the check for you and send signatured information regarding to browser client

#### Apache2

Add the following line INSIDE the `` <VirtualHost></VirtualHost> `` block:
```
SSLUseStapling on
SSLStaplingResponderTimeout 5
SSLStaplingReturnResponderErrors off
```
Add the following line OUTSIDE  the ``` <VirtualHost></VirtualHost> ``` block:
```
SSLStaplingCache shmcb:/tmp/stapling_cache(128000)
```

#### NGINX

```
    ssl_stapling on;
    ssl_stapling_verify on;
    ssl_stapling_file ocsp_response;
    ssl_trusted_certificate /path/to/cert_chain.pem;

    resolver <IP DNS resolver>;
```

#### Check the OCSP configuration

```

> openssl s_client -connect yourservername.org:443 -servername yourservername.org -status < /dev/null 2>&1 | grep -i "OCSP response"

Reviewer 3: Maybe add the expected output?

Author: good suggestion, i will add this to the document.

This protects e.g. from protocol downgrade attacks and cookie hijacking.

By adding the following to your SSL VirtualHost configuration you instruct browsers not to allow any connection to your ILIAS instance using HTTP and prevent visitors from bypassing invalid certificate warnings.

#### Apache2

Ensure you have ```mod_headers.so``` enabled in Apache2:

```
<IfModule mod_headers.c>
    Header always set Strict-Transport-Security "max-age=15552000; includeSubDomains; preload"
</IfModule>
```

#### NGINX

```
    add_header Strict-Transport-Security max-age=15552000; includeSubDomains" preload;
```

**Warning:** Before activating the configuration above you MUST make sure that you have a good workflow for maintaining your SSL settings (including certificate renewals) as you will not be able to disable HTTPS access to your site for up to 6 months.

**Tip:** When you test HSTS, use a very short max-age timeout
```
includeSubDomains:       Restrictions also apply to all subdomains of the current domain.
max-age:                 Duration of cached information (180 days)
```

### Genrate the ILIAS ```ilClientId``` cookie with ```secure``` attribute

> The browser will include the cookie in an HTTP request only if the request is transmitted over a secure channel (HTTPS).  

Reviewer 3: I'm not sure if that is needed anymore. The `Secure` flag seems to be set already by default, at least when HTTPS is used. I'm not sure if it is the browser, the webserver or ILIAS itself that is enforcing this behavior, though.

Author: A JF decision (https://docu.ilias.de/goto.php?target=wiki_1357_JourFixe-2018-04-23) was to not send the secure flag during initialization of ILIAS.

The workshop discussed to add the flag via web server until ILIAS will send this.

```
    more_set_headers 'Server: ';
```
or better
```
    more_clear_headers  'Server:*';
```

### Remove “X-Powered-By” information from http header

This can also be done by unsetting the header in webserver.
* "Header unset X-Powered-By" for Apache2 or rather
* "more_clear_headers 'X-Powered-By'" for NGINX.

The suggested solution is to set 'expose_php' to 'off' in your php.ini:

```
    expose_php = Off
```

## deny access or restrict to several files or locations

### ILIAS setup
The access to the ILIAS Installation Wizard ``(/setup/setup.php)`` MAY be restricted:<br/><br/>

Apache2:  
```
    <Location /setup>
        <IfVersion < 2.3>
         Order Deny,Allow
         Deny From All
         Allow from %IP-Address
        </IfVersion>

        <IfVersion > 2.3>
            Require all denied
            Require ip %IP-Address
        </IfVersion>
    </Location>
```

Nginx:  

```
    location /setup {
        allow %IP-Address;
        deny all;

        # add location for PHP processing here
    }
    location /setup/setup.php {
        allow %IP-Address;
        deny all;
    }

```

### Prevent blacklisted files of beeing served by the webserver

> If somebody tries to upload a file thats filetype is blacklisted by the upload settings, the upload will take place but the file will be renamed to ``filename.sec``. The webserver MUST not serve this file anymore to it's visitors as the file may consists of maliscious software.

Reviewer 3: I don't think this is a recommended practice and I strongly disagree. If files with unknown file types are not to be served, then the application itself should deny the download, not the web server.

Author: There is a heavy discussion about this.

In workshop we discussed to change "MUST" to "should"....

Please note that this will not work with Nginx as ``.htaccess-files`` are not supported. Instead you MUST add the following to your Nginx configuration file (please note that running ILIAS with Nginx isn't officially supported and certain features like Shibboleth won't work):

This is a NGINX recommended configuration. (note: inside the `%docroot_dir%/data` no PHP will be proceed)

```
    server {
        [...]
        location ~ /data/ {
            set $root $document_root;
            rewrite ^/data/(.*)/(.*)/(.*)$ /Services/WebAccessChecker/wac.php last;

            location ~ [^/]\.php(/|$) { access_log off; log_not_found off; deny all; }
        }

        location ^~ /secured-data {
            # Protected, only working for subrequests (X-Accel-Redirect)
            alias $root/data;
            internal;

            location ~ [^/]\.php(/|$) { access_log off; log_not_found off; deny all; }
        }
         [...]
    }
```

> ** [ilFileDelivery](https://github.com/ILIAS-eLearning/ILIAS/blob/release_5-4/Services/FileDelivery/classes/override.php.template)** (concerns NGINX/PHP-FPM):

Reviewer: What is "ilFileDelivery" and why would I use it? Which implications, if any, might be noteworthy?

Author: ilFileDelivery is also described in the template files itself.

For a good performance, it is suggested.
OK, this is also more a performance improvement.

> This is needed if you want to use the ilFileDelivery::DELIVERY_METHOD_XACCEL or the ilFileDelivery::DELIVERY_METHOD_XSENDFILE Method since PHP can't figure out whether X-Accel ist installed or not.""

rename the file:  
     ```%docroot_dir%/Services/FileDelivery/classes/override.php.template```  
to  
     ```%docroot_dir%Services/FileDelivery/classes/override.php```  

## Use secure passwords
Please keep in mind, that your plattform might me be accessible to the world wide web. To avoid unauthorized access to your Ilias-Installation, it his highly recommended to use secure passwords. Especially the root password and the Ilias-Master-Password are potentially endangered.

Your passwords should fullfil the following criterias:

* at least 8 characters in length

> * lowercase and uppercase alphabetic characters, numbers and symbols

Reviewer 3: ILIAS does not accept some common symbols. Maybe we could be more specific here?

Author: Good suggestion! We add this to the document.

    + [Prevent blacklisted files of beeing served by the webserver](#prevent-blacklisted-files-of-beeing-served-by-the-webserver)
    + [Prevent execution of PHP-Code in data-directory](#prevent-execution-of-php-code-in-data-directory)
    + [Deny Access to local Git-Directory](#deny-access-to-local-git-directory)
  * [Integrity check of ILIAS code in docroot](#integrity-check-of-ilias-code-in-docroot)
  * [Use WebAccessChecker](#use-webaccesschecker)
  * [Use secure passwords](#use-secure-passwords)
  * [Report security issues](#report-security-issues)

# Hardening and Security Guidance

This guideline is going to show you some best practice examples how to improve the security of your ILIAS installation, of the webserver and it's associated components.
The ILIAS e.V. requested a documentation for a "Secure ILIAS" in march 2019.
The section "Hardening and Security Guidance" should be removed and the security related instructions have to be mantained in a own document.

## List of placeholder names

In the following text includes some placeholder.
For a better identification, they will describe here:

|  placeholder |  description |
|-------|-------|

> |  %USERNAME% |  a choiceable user name or the webserver user based on distribution  |

Reviewer 5: The correct word would be "choosable", although I would suggest to use individual
```suggestion
| %USERNAME% |  an individual user name or the webserver user based on distribution  |
```

> | %GROUP% | a choiceable group name or the webserver group based on distribution|

Reviewer 5: The correct word would be "choosable", although I would suggest to use individual
```suggestion
| %GROUP% | an individual group name or the webserver group based on distribution|
```

    ssl_prefer_server_ciphers on;

    ssl_dhparam /etc/ssl/private/dhparam.pem;

```

``ssl_dhparam /etc/ssl/private/dhparam.pem;``:

This specifies a file with DH parameters for EDH (Ephemeral Diffie-Hellman) ciphers.  
By default, NGINX will use the default DHE paramaters provided by openssl. This uses a weak key that gets lower scores.  
Run ``openssl dhparam -out /etc/ssl/private/dhparam.pem 4096`` in terminal to generate it.

We RECOMMEND to use the [Mozilla SSL Configuration Generator](https://ssl-config.mozilla.org/) to generate a suitable configuration and the [Qualys SSL Labs Tests](https://www.ssllabs.com/ssltest/) or the [High-Tech Bridge SSL Server Test](https://www.htbridge.com/ssl/) to check your settings. It is recommended, to reach a "A" rating as minimum.

It is necessary to often revise these configuration.
In best case, you allways use the latest "Modern" configuration.

### Serve security related Headers

To improve the security of your ILIAS users you SHOULD set the following Headers:

> (this is experience based configuration set of headers, may this can differ to your configuration/ usage scenario)

Reviewer 5: Little bit of English
```suggestion
(this is an experience based configuration set of headers, this may differ from your configuration/usage scenario)
```

         add_header X-Content-Security-Policy "default-src 'self'; connect-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval' blob:; style-src 'self' 'unsafe-inline' data:; img-src 'self' 'unsafe-inline' data:; font-src 'self' 'unsafe-inline' data:; media-src 'self' 'unsafe-inline' data:";
 ```

note:  
If you use a proxied [Chat Server](https://github.com/ILIAS-eLearning/ILIAS/blob/release_5-4/Modules/Chatroom/README.md), you MUST add the url to the CSP definition:  
```connect-src 'self' wss://onscreenchat.%HOSTNAME% https://onscreenchat.%HOSTNAME%;```

If you use other external content (e.g. for a SCORM module or external video content) you have to add a ```default-src``` (The default-src is the default policy for loading content such as JavaScript, Images, CSS, Fonts, AJAX requests, Frames, HTML5 Media from other sites):

e.g. for https://www.youtube.com as source:

```default-src 	'self' www.youtube.com; ...```

#### Validate the header configuration
It is recommended to validate your configuration with the services from https://securityheaders.com/.  
Try to reach **A** grade.

### Enable OCSP stapling (TLS Certificate Status Request)

> The  TLS/SSL extension [OCSP stapling](https://en.wikipedia.org/wiki/OCSP_stapling) provides a way to improve the performance (cause the webserver is will do this instead of the browser client) of SSL negotiation while maintaining visitor privacy.

Reviewer 5: ```suggestion
The  TLS/SSL extension [OCSP stapling](https://en.wikipedia.org/wiki/OCSP_stapling) provides a way to improve the performance (as the webserver will fetch the OCS instead of the browser client) of SSL negotiation while maintaining visitor privacy.
```

By adding the following to your SSL VirtualHost configuration your webserver will do the check for you and send signatured information regarding to browser client

#### Apache2

Add the following line INSIDE the `` <VirtualHost></VirtualHost> `` block:
```
SSLUseStapling on
SSLStaplingResponderTimeout 5
SSLStaplingReturnResponderErrors off
```
Add the following line OUTSIDE  the ``` <VirtualHost></VirtualHost> ``` block:
```
SSLStaplingCache shmcb:/tmp/stapling_cache(128000)
```

#### NGINX

```
    ssl_stapling on;
    ssl_stapling_verify on;
    ssl_stapling_file ocsp_response;
    ssl_trusted_certificate /path/to/cert_chain.pem;

    resolver <IP DNS resolver>;
```

#### Check the OCSP configuration

```
openssl s_client -connect %HOSTNAME%:443 -servername %HOSTNAME% -status < /dev/null 2>&1 | grep -i "OCSP response"
```

The expected output of a working OCSP configuration:
```
OCSP response:
OCSP Response Data:
    OCSP Response Status: successful (0x0)
    Response Type: Basic OCSP Response
```

> The [Qualys SSL Labs Tests](https://www.ssllabs.com/ssltest/) will also provided you a working or non-working OCSP configuration:

Reviewer 5: ```suggestion
The [Qualys SSL Labs Tests](https://www.ssllabs.com/ssltest/) will also show if your OCSP configuration is working:
```

</IfModule>
```

#### NGINX
* Module ngx_http_headers_module is needed

```
    more_set_headers 'Server: ';
```
or better
```
    more_clear_headers  'Server:*';
```

### Remove “X-Powered-By” information from http header

This can also be done by unsetting the header in webserver.
* "Header unset X-Powered-By" for Apache2 or rather
* "more_clear_headers 'X-Powered-By'" for NGINX.

The suggested solution is to set 'expose_php' to 'off' in your php.ini:

```
    expose_php = Off
```

## deny access or restrict to several files or locations

### ILIAS setup
The access to the ILIAS Installation Wizard ``(/setup/setup.php)`` MAY be restricted:<br/><br/>

Apache2:  
```
    <Location /setup>
        <IfVersion < 2.3>
         Order Deny,Allow
         Deny From All
         Allow from %IPADDRESS%
        </IfVersion>

        <IfVersion > 2.3>
            Require all denied
            Require ip %IPADDRESS%
        </IfVersion>
    </Location>
```

Nginx:  

```
    location /setup {
        allow %IPADDRESS%;
        deny all;

        # add location for PHP processing here
    }
    location /setup/setup.php {
        allow %IPADDRESS%;
        deny all;
    }

```

> Please add here the whitelisted ip address (%IPADDRESS%) to grant access to ILIAS setup here.

Reviewer 5: ```suggestion
Please add the whitelisted ip address (%IPADDRESS%) to grant access to ILIAS setup here.
```

### Prevent blacklisted files of beeing served by the webserver

> If somebody tries to upload a file thats filetype is blacklisted by the upload settings, the upload will take place but the file will be renamed to ``filename.sec``. The webserver should not serve this file anymore to it's visitors as the file may consists of maliscious software.

Reviewer 5: ```suggestion
If somebody tries to upload a file with a filetype blacklisted by the upload settings, the upload will take place, but the file will be renamed to ``filename.sec``. The webserver should not serve this file anymore to it's visitors as the file may consists of malicious software.
```
